lib/dispatch-rider/publisher/base.rb

```diff
@@ -1,37 +1,46 @@
 require 'securerandom'
 
 module DispatchRider
+  # Main template for a dispatch rider publisher.
   class Publisher::Base
+    include Callbacks::Support
 
     class << self
-
+      # @param [Symbol] subject
       def subject(subject)
         @subject = subject
       end
 
+      # @param [Array<Symbol>, Symbol] destinations
       def destinations(destinations)
         @destinations = Array(destinations)
       end
 
+      # @return [DispatchRider::Publisher]
       def default_publisher
         @@default_publisher ||= DispatchRider::Publisher.new
       end
 
       def publish(*args, &block)
-        raise NotImplementedError
+        raise NotImplementedError, "subclass of DispatchRider::Publisher::Base must implement .publish"
       end
-
     end
 
     def initialize(publisher = nil)
       @publisher = publisher
     end
 
+    # @param [Hash] body
     def publish(body)
       raise ArgumentError, 'body should be a hash' unless body.kind_of?(Hash)
-      publisher.publish(destinations: destinations, message: { subject: subject, body: body })
+
+      callbacks.invoke(:publish, body) do
+        publisher.publish(destinations: destinations, message: { subject: subject, body: body })
```

Comment:
Hmmm... You're right we're no longer following the prescribed way from the older time.

Moving this to `DispatchRider::Publisher` itself then and not the base template.